engine/src/main/scala/cromwell/engine/db/slick/SlickDataAccess.scala

```diff
@@ -181,16 +180,39 @@ class SlickDataAccess(databaseConfig: Config) extends DataAccess {
     val actionThreadPoolSize = databaseConfig.getIntOr("actionThreadPoolSize", dbNumThreads) min dbMaximumPoolSize
     Executors.newFixedThreadPool(actionThreadPoolSize)
   }
+
   private val actionExecutionContext: ExecutionContext = ExecutionContext.fromExecutor(
-    actionThreadPool, database.executor.executionContext.reportFailure)
+    actionThreadPool, database.executor.executionContext.reportFailure
+  )
+
+  private lazy val useSlickUpserts = dataAccess.driver.capabilities.contains(JdbcProfile.capabilities.insertOrUpdate)
 
   override def close(): Unit = {
     actionThreadPool.shutdown()
     database.close()
   }
 
-  // Run action with an outer transaction
-  private def runTransaction[R](action: DBIOAction[R, _ <: NoStream, _ <: Effect]): Future[R] = {
+  private def runTransactionWithRollbackRetry[R](action: DBIOAction[R, _ <: NoStream, _ <: Effect], actorSystem: ActorSystem): Future[R] = {
+    implicit val ec = actorSystem.dispatcher
+    def retry[T](f: => Future[T], delay: FiniteDuration, retries: Int): Future[T] = {
```

Comment:
Regarding our chat on an explicit execution context: I would also be ok if you swapped the above two lines.

The swap would then scope the `implict ec` **_only_** within the `def retry {}` block. I don't want any implicit execution contexts floating around within the scope of the wider `def runTransactionWithRollbackRetry {}` block.